NetDb: refactor getting closest routers

coneiric opened the pull request:
---
**By submitting this pull-request, I confirm the following:**

- I have read and understood the developer guide in [kovri-docs](https://github.com/monero-project/kovri-docs).
- I have checked that another pull-request for this purpose does not exist.
- I have considered and confirmed that this submission will be valuable to others.
- I accept that this submission may not be used and that this pull-request may be closed by the will of the maintainer.
- I give this submission freely under the BSD 3-clause license.
---

A number of refactors to getting closest routers:

* Adds error handling for `CreateRoutingKey` callers.
* Excludes routers before considering them as closest candidates.
* Adds lock for `RouterInfo`s in `GetClosestNonFloodfill` to make calling the function thread-safe.
* Unit-tests for the refactored functions.

Referencing #643.

anonimal commented on src/core/router/net_db/impl.cc:
```diff
@@ -978,29 +979,23 @@ std::vector<IdentHash> NetDb::GetClosestFloodfills(
-    if (i < num) {
-      auto& ident = it.r->GetIdentHash();
-      if (!excluded.count(ident)) {
+    while (i < num)
```
This will insert the same hash repeatedly... and any remaining elements in the set will be ignored.

coneiric replied:
You're right, I'll change the loop to step through the whole set.

anonimal commented on src/core/router/net_db/impl.cc:
```diff
@@ -978,29 +979,23 @@ std::vector<IdentHash> NetDb::GetClosestFloodfills(
+
   std::vector<IdentHash> res;
   std::uint8_t i{};
   for (auto it : sorted) {
```
May as well `const auto&` while you're here.

coneiric replied:
Will do.

anonimal commented on src/core/router/net_db/impl.cc:
```diff
@@ -978,29 +979,23 @@ std::vector<IdentHash> NetDb::GetClosestFloodfills(
   IdentHash dest_key = CreateRoutingKey(destination); {
     std::unique_lock<std::mutex> l(m_FloodfillsMutex);
     for (auto it : m_Floodfills) {
```
May as well `const auto&` while you're here.

coneiric replied:
Will do.

anonimal commented on src/core/router/net_db/impl.cc:
```diff
@@ -951,13 +951,14 @@ std::shared_ptr<const RouterInfo> NetDb::GetClosestFloodfill(
   min_metric.SetMax();
   std::unique_lock<std::mutex> l(m_FloodfillsMutex);
   for (auto it : m_Floodfills) {
```
May as well `const auto&` while you're here.

coneiric replied:
Will do.

anonimal commented on src/core/router/net_db/impl.h:
```diff
@@ -197,16 +197,16 @@ class NetDb : public NetDbTraits {
   std::shared_ptr<const RouterInfo> GetClosestFloodfill(
       const IdentHash& destination,
-      const std::set<IdentHash>& excluded) const;
+      const std::set<IdentHash>& excluded);
```
These getters should remain const. They should only be accessing members, not mutating.

coneiric replied:
> These getters should remain const

I thought so as well, but didn't know how much work would be needed on the dispatcher to make it const as well. Will work on this.

anonimal replied:
Not much to it:

```diff
diff --git a/src/core/util/exception.cc b/src/core/util/exception.cc
index 936dfa9f..4bc5ad23 100644
--- a/src/core/util/exception.cc
+++ b/src/core/util/exception.cc
@@ -45,7 +45,7 @@ namespace core {
 Exception::Exception(const char* message) : m_Message(message) {}
 
 // TODO(anonimal): exception error codes to replace strings?
-void Exception::Dispatch(const char* message)
+void Exception::Dispatch(const char* message) const
 {
   // Message to log
   std::string log;
diff --git a/src/core/util/exception.h b/src/core/util/exception.h
index dc84d926..e7ae5cc1 100644
--- a/src/core/util/exception.h
+++ b/src/core/util/exception.h
@@ -49,7 +49,7 @@ class Exception final {
   /// @brief Exception class dispatcher
   /// @details Set optional exception message, concats messages, adds trivial formatting
   /// @param message String message to log for exception
-  void Dispatch(const char* message = "");
+  void Dispatch(const char* message = "") const;
 
  private:
   std::string m_Message;
```

coneiric replied:
Wow, really not much. Will make the changes.

anonimal commented on src/core/router/net_db/impl.h:
```diff
@@ -197,16 +197,16 @@ class NetDb : public NetDbTraits {
       const IdentHash& destination,
       std::uint8_t num,
-      std::set<IdentHash>& excluded) const;
+      std::set<IdentHash>& excluded);
```
These getters should remain const. They should only be accessing members, not mutating.

coneiric replied:
> These getters should remain const

I thought so as well, but didn't know how much work would be needed on the dispatcher to make it const as well. Will work on this.

anonimal commented on src/core/router/net_db/impl.h:
```diff
@@ -197,16 +197,16 @@ class NetDb : public NetDbTraits {
   std::shared_ptr<const RouterInfo> GetClosestNonFloodfill(
       const IdentHash& destination,
-      const std::set<IdentHash>& excluded) const;
+      const std::set<IdentHash>& excluded);
```
These getters should remain const. They should only be accessing members, not mutating.

coneiric replied:
> These getters should remain const

I thought so as well, but didn't know how much work would be needed on the dispatcher to make it const as well. Will work on this.

anonimal commented on tests/unit_tests/core/router/net_db/impl.cc:
```diff
@@ -0,0 +1,183 @@
+{
+  /// @alias RouterCaps
+  /// @brief Capabilities alias for readability and convenience
+  using RouterCaps = kovri::core::RouterInfoTraits::Cap;
```
To keep in line with my branch 'bandcaps', s/RouterCaps/Cap/ (this will speed up refactoring once that's merged).

coneiric replied:
Will do, meant to change before pushing.

coneiric commented:
> These tests are a perfect example of why we simply need to test our kademlia implementation.

I can work on this as well, and will read up on Kademlia specs.

> please condense all this copypasta into reusable functions and/or a fixture

Will do, and I originally had more of the code in the fixture. The noise from `RouterInfo` logging made the test logs nearly unreadable. I'll keep working on a good test fixture.

coneiric commented:
Just added the recommended changes, and updated the test fixture. If the changes look good, I'll add the issue + PR references to the commit messages.

coneiric commented:
> Can you please apply the following and add a "credit to anonimal" in the git log?

Absolutely, my bad for not including the credit for the last patch you submitted to my other PR.

Thank you for the patch.
